### PR TITLE
Update nn.py

### PR DESCRIPTION
**Pull request by pctablet505**
Added support for flash attention with sharding, fixed issue when using flash attention on tpu.

### CODE REVIEW

File: keras/src/backend/jax/nn.py [hunk 1]
```diff
@@ -1177,54 +1178,111 @@ def dot_product_attention(
...
-            custom_mask=mask,
-            attn_logits_soft_cap=attn_logits_soft_cap,
+    # Check if inputs are sharded
+    inputs_sharded = any(
+        hasattr(t, "sharding") and not t.sharding.is_fully_replicated
+        for t in (query, key, value)
+    )
+    is_tpu = jax.devices()[0].platform == "tpu"
+
+    # Determine flash attention compatibility
```
Comment (divyashreepathihalli) [inline keras/src/backend/jax/nn.py:1189]: I am very confused about the logic here.

- why is FA disabled in inputs are sharded?

File: keras/src/backend/jax/nn.py [hunk 2]
```diff
@@ -1177,54 +1178,111 @@ def dot_product_attention(
...
+    if flash_attention is None:
+        flash_attention = (
+            not inputs_sharded or is_tpu
+        ) and _can_use_flash_attention(query, key, value, bias)
+    elif flash_attention and inputs_sharded and not is_tpu:
```
Comment (divyashreepathihalli) [inline keras/src/backend/jax/nn.py:1194]: this condition is weird
if FA is enabled and inputs are sharded and is not running on TPU - you are disabling FA? why? can you please explain?

File: keras/src/backend/jax/nn.py [hunk 3]
```diff
@@ -1177,54 +1178,111 @@ def dot_product_attention(
...
+        segment_ids = jnp.zeros([bs, q_len], dtype=jnp.int32)
+        decoder_segment_ids = splash_attention_kernel.SegmentIds(
+            q=segment_ids, kv=segment_ids
         )
-        out = jnp.transpose(out, axes=(0, 2, 1, 3))
-        return out
 
-    # `dot_product_attention` is only available in jax>=0.4.31
+        # Process mask for Splash Attention
+        custom_mask = None
```
Comment (divyashreepathihalli) [inline keras/src/backend/jax/nn.py:1219]: lets verify numerics remain consistent with this updated code to mask

**Review by divyashreepathihalli — COMMENTED** [review]
Thanks for the PR!
The gating logic is a little confusing for me. I left some comments. Thanks!
